recipes/hdf5/all/conanfile.py

```diff
@@ -106,28 +107,26 @@ def _cmake_new_enough(self, required_version):
         except:
             return False
 
-    def _inject_stdlib_flag(self, tc):
-        if self.settings.os == "Linux" and self.settings.compiler == "clang":
-            cpp_stdlib = f" -stdlib={self.settings.compiler.libcxx}".rstrip("1")  # strip 11 from stdlibc++11
-            tc.variables["CMAKE_CXX_FLAGS"] = tc.variables.get("CMAKE_CXX_FLAGS", "") + cpp_stdlib
-
-        return tc
-
-
     def build_requirements(self):
         if Version(self.version) >= "1.14.0" and not self._cmake_new_enough("3.18"):
-            self.tool_requires("cmake/3.25.0")
+            self.tool_requires("cmake/3.25.3")
 
     def source(self):
         get(self, **self.conan_data["sources"][self.version], strip_root=True)
 
+    def _inject_stdlib_flag(self, tc):
+        if self.settings.os == "Linux" and self.settings.compiler == "clang":
+            cpp_stdlib = f" -stdlib={self.settings.compiler.libcxx}".rstrip("1")  # strip 11 from stdlibc++11
+            tc.variables["CMAKE_CXX_FLAGS"] = tc.variables.get("CMAKE_CXX_FLAGS", "") + cpp_stdlib
+        return tc
+
     def generate(self):
         cmakedeps = CMakeDeps(self)
         cmakedeps.generate()
 
         tc = CMakeToolchain(self)
-        if self.settings.get_safe("compiler.cppstd"):
-            tc.variables["CMAKE_CXX_STANDARD"] = self._min_cpp_standard
+        if not valid_min_cppstd(self, self._min_cppstd):
+            tc.variables["CMAKE_CXX_STANDARD"] = self._min_cppstd
```

Comment:
I realise that this is just a refactor of existing logic, but is this really desirable behaviour? The way I'm reading this is that if a user specifies the following profile trying to build > 1.14
```
[settings]
compiler.cppstd=03
```
Then this would say "nah, actually i'll build with `compiler.cppstd=11`". That's pretty unintuitive and probably has ABI implications. I think this is probably better placed in `validate()` to just throw an error if user requests a `cppstd` that isn't supported rather than trying to auto-fix the error for them

I think this kind of logic would only be appropriate when testing for `cppstd=None`, so something closer to the previous logic:

```python
if not self.settings.get_safe("compiler.cppstd"):
    tc.variables["CMAKE_CXX_STANDARD"] = self._min_cppstd
```

Reply:
With the logic you propose, default cppstd of compiler may not be honored when self.settings.compiler.cppstd is not set (which is allowed in conan v1, and still technically allowed in conan v2).

> Then this would say "nah, actually i'll build with compiler.cppstd=11". That's pretty unintuitive and probably has ABI implications.

It can't happen. If someone set `compiler.cppstd=03`, it will raise in `validate()`.

Here are different scenario of current logic:

- compiler.cppstd is set:
  - if it's less than self._min_cppstd, it raises
  - if it's greater than self._min_cppst, CMakeToolchain injects CMAKE_CXX_STANDARD accordingly and everything works as user expects
- if compiler.cppstd is not set:
  - if default compiler C++ standard is less than self._min_cppstd, it forces CMAKE_CXX_STANDARD to self._min_cppstd, which is more a less what would have happened if upstream CMakeLists was properly writtent with target_compile_features
  - if default compiler C++ standard is greater than self._min_cppstd, this recipe doesn't inject CMAKE_CXX_STANDARD, and it works fine as well.

Reply:
Yep that makes sense, I'm happy with that - I wasn't looking at the implementation of `validate()` when I wrote this. Thanks for the explanation